ctk/common/src/main/kotlin/org/codice/compliance/utils/XMLDecryptor.kt

```diff
@@ -0,0 +1,163 @@
+                    .getProperty("org.apache.ws.security.crypto.merlin.keystore.alias")
+
+            // Using Apache Merlin to read the keystore
+            val merlin = Merlin(encryptionProperties,
+                    ClassLoader.getSystemClassLoader(),
+                    JasyptPasswordEncryptor(keystorePassword))
+
+            merlin.getPrivateKey(keystorePrivateKeyAlias, keystorePrivateKeyPassword)
+        }
+
+        /**
+         * Takes in an EncryptedID, EncryptedAssertion, or EncryptedAttribute element and returns
+         * the decrypted node.
+         *
+         * Only supports up to one level encryption using the EncryptedKey element. In other words,
+         * the data must be directly encrypted by the CTK's public key, or, if another key is used
+         * to encrypt the data, that key must be encrypted using the CTK's public key and put into
+         * the EncryptedData element.
+         *
+         * @param node An encrypted {@code Node} to be decrypted
+         * @return Node The decrypted {@code Node}
+         */
+        @Suppress("TooGenericExceptionCaught"
+                /* Decryption error handling is independent of Exception type */)
+        fun decryptAndReplaceNode(node: Node): Node {
+            // message to build up and pass at the end of the method
+            val messageBuilder = StringBuilder()
+
+            val encElement = node as Element
+            val ownerDocument = node.ownerDocument
+            val encData = encElement.children("EncryptedData").first() as Element
+
+            val encKeyElements = encElement.allChildren("EncryptedKey")
+                    .map { it as Element }
+            if (encKeyElements.isEmpty()) messageBuilder.append("No Keys could be found.\n")
+
+            // If given KeyInfo, decrypt said key and use it. Else use the default server key.
+            val encryptionKey = tryKeyDecryption(encKeyElements, ownerDocument)
+                    ?: serverPrivateKey
+            if (encryptionKey == serverPrivateKey)
+                messageBuilder.append("No Keys could be decrypted.\n")
+
+            val encryptionCipher = XMLCipher.getProviderInstance(BC_PROVIDER).apply {
+                init(XMLCipher.DECRYPT_MODE, encryptionKey)
+            }
+
+            try {
+                // Keep a reference Node, either a sibling (if it has one) or parent, of `node` in
+                // order to return the decrypted version of `node`
+                return if (encData.previousSibling != null) {
+                    val referenceNode = encData.previousSibling
+                    encryptionCipher.doFinal(ownerDocument, encData)
+                    replaceNode(referenceNode.nextSibling, node)
+                } else {
+                    val referenceNode = encData.parentNode
+                    encryptionCipher.doFinal(ownerDocument, encData)
+                    replaceNode(referenceNode.firstChild, node)
```

Comment:
✏️ I still don't like that you're doing this weird replace based on either the sibling or the parent. I liked the way Keycloak was doing it with using the returned document from `encryptionCipher.doFinal()` and replacing from the root node.

Reply:
👍  I have an idea that may make it look cleaner. If that ends up being a bust I'll look back at Keycloak's solution.

Reply:
My idea didn't plan out. And as for Keycloak, they are making certain assumptions that we can't. They assume that the first child of the root element is the encrypted element. I'm not sure if this is because they put the encrypted element into a new document before sending it to the XML processing function, or if they just make bad assumptions, but either way we can't assume that for our purposes. Any other ideas are welcome though!

Keycloack reference: https://github.com/keycloak/keycloak/blob/17c13043d0118ecb866c294e32c30a7bcb8ddb13/saml-core/src/main/java/org/keycloak/saml/processing/core/util/XMLEncryptionUtil.java#L378

Reply:
Yeah you're right. They create a brand new document and pass that into the function... I don't have any better ideas.